enhancements/results-forwarding.md

```diff
@@ -0,0 +1,504 @@
+to a configured implementation. In this case, the **aggregator** will act as a
+gRPC client. The **resultserver** then becomes a gRPC server, subject to the
+evidence forwarding provider and will be renamed **evidence-persistor** and
+responsible for writing results.
+
+Forwarder implementations would be configured via the `ScanSetting` and they'd be called "providers".
+For backwards compatibility, we'd have defaults that would point to the current mode of operation
+of the Compliance Operator. e.g. a `pvc` provider for evidence storage.
+
+For more detail on the proposed API, see the *API Extensions* section.
+
+### User Stories
+
+* As an Site Reliability Engineer managing multiple clusters, I'd like to run
+  compliance scans on my fleet in such a way that uses the minimal amount of
+  resources.
+
+* As a Site Reliability Engineer managing multiple clusters, I'd like to have
+  one place to view the compliance stance of my whole fleet.
+
+### API Extensions
+
+Given the scope of the change, we'd upgrade the `ScanSetting` version to
+`v1alpha2`. The following sections describe the additional values needed by the
+`ScanSetting` CRD.
+
+#### `resultForwarding.provider`
+
+Required: True
+Type: string
+
+The implementation to use for forwarding. The initial implementation will
+support `grpc`.
+
+#### `resultForwarding.grpc.endpoint`
+
+Required: True
+Type: string
+
+Endpoint URL to a gRPC server for result storage. The service may be in or
+outside the cluster. One potential implementation may be
+[compserv](https://github.com/rhmdnd/compserv).
+
+#### `resultForwarding.grpc.version`
+
+Required: False
+Default: `v1`
+Type: string
+
+#### `resultForwarding.grpc.tls.dynamic.caSecretName`
+
+Required: True
+Type: string
+
+Reference to a CA certificate and key pair for TLS.
+
+#### `resultForwarding.grpc.tls.dynamic.serverSecretName`
+
+Required: True
+Type: string
+
+Reference to a server certificate/keypair secret for mutual TLS.
+
+#### `resultForwarding.grpc.tls.dynamic.clientSecretName`
+
+Required: True if using mutual TLS
+Type: string
+
+Reference to a client certificate/keypair secret for mutual TLS.
+
+#### `resultForwarding.grpc.authentication.type`
+
+Required: False
+Default: `mtls`
+Type: string
+Choices: `mtls`, `token`
+
+The type of authentication the Compliance Operator should use to authenticate
+the forwarding endpoint.
+
+#### `resultForwarding.grpc.authentication.token`
+
+Required: True if using `token` authentication type
+Type: string
+
+#### `resultForwarding.grpc.authentication.token.tokenSecretName`
+
+Required: False
+Type: string
+
+Reference to the secret containing the token.
+
+#### `resultForwarding.grpc.extraMetadata.clusterName`
+
+Required: True
+Type: string
+
+The name of the cluster to use in result payloads.
+
+#### `resultForwarding.grpc.extraMetadata.clusterType`
+
+Required: True
+Type: string
+
+The type of the cluster to use in result payloads.
```

Comment:
I think I may have typo'd this.  What would be an example of different cluster types? Or was this supposed to be the difference between targets (e.g., running a compliance scan on a `server` versus a `cluster` where `server` and `cluster` are two different types)?

Reply:
maybe the type can be ex. `metal`, `aws`.`gcp`? If that's the case, we should be able to derive this from cluster api

Reply:
Ok - sounds reasonable to keep it `clusterType`. I think I got my wires crossed thinking about this and the compliance service schema. Since this forwarding implementation is Kubernetes specific, then keeping the name `clusterType` makes sense.

Reply:
I think we can ditch this from the definition, but it would still be nice to have "extras" as something we can extend and add random attributes. We have types of clusters based on confidentiality levels, and there are types just based on usage. e.g. there's a type cluster for certain types of applications and another type of cluster for another set of applications.

Reply:
So remove this from the `extraMetadata` and add it when it's needed? Thinking about this from the compserv perspective, I'm wondering how this would be useful there (would it be used as a filter?)

Reply:
if you want to have extra key-value pairs sent, then why not have a reference to a secret or a CM and send along its keys/values?

Reply:
Yeah - that could work.

The thing I'm curious about here is if we should allow a generic mechanism for forwarding extra data, or be more strict about it.

Reply:
hmm, maybe? I can see it as useful, the only thing we'd have to be careful about is to not abuse the extra fields later on (as opposed to changing the grpc protocol itself)